lib/fluent/plugin/parser.rb

```diff
@@ -17,13 +17,58 @@
 require 'fluent/plugin/base'
 require 'fluent/plugin/owned_by_mixin'
 
+require 'fluent/error'
 require 'fluent/mixin' # for TypeConverter
 require 'fluent/time'
 require 'fluent/plugin/string_util'
 
+require 'serverengine/blocking_flag'
+
 module Fluent
   module Plugin
     class Parser < Base
+      class TimeoutChecker
+        # This implementation now uses mutex because parser is typically used in input.
+        # If this has a performance issue under high concurreny, use concurrent-ruby's map instead.
+        def initialize(timeout)
+          @map = {}
+          @flag = ServerEngine::BlockingFlag.new
+          @mutex = Mutex.new
+          @timeout = timeout
+        end
+
+        def start
+          @thread = ::Thread.new {
+            until @flag.wait_for_set(0.5)
```

Comment:
In this implementation, it can detect Timeout only 0.5-sec presicion.  Is 0.25 better than 0.5 since logstash did
https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html#plugins-filters-grok-timeout_millis

Reply:
I'm not sure which is the best because ruby and java are different VM.
Frequent mutex / cond access is also heavy task. In addition, very small timeout is not realistic because timeout precision is not exact for smaller value. Rough check seems reasonable for this case.
One idea is make timeout check parameterized.

Reply:
>  Rough check seems reasonable for this case.

I see. I think it's better to add the document(or other things users can access) about timeout precision.